### PR TITLE
fix(core): Fix arguments parsing method of `Invoke-ExternalCommand()`

### PR DESCRIPTION
**Pull request by niheaven**
#### Description

Final PR for stable version (#5424)

Before handling the argument list, it splits by finding spaces that are not in a path or before params (starting with `/` or `-`) and adding them to `Process.StartInfo.ArgumentList` or `Process.StartInfo.Arguments`.

Example: (https://regex101.com/r/HynlRG/1)

image

#### Motivation and Context

- Relates to #5065 

#### How Has This Been Tested?

Test with the following manifests, and scoop path with/without spaces.

<details>
<summary>msmpi</summary>

```json
{
    "version": "10.1.1",
    "description": "Microsoft implementation of the Message Passing Interface standard for developing and running parallel applications on the Windows platform.",
    "homepage": "https://docs.microsoft.com/en-us/message-passing-interface/microsoft-mpi",
    "license": "MIT",
    "url": "https://github.com/Microsoft/Microsoft-MPI/releases/download/v10.1.1/msmpisetup.exe",
    "hash": "7308fd15e437d6829fd9c6ec5d801380faa3bff6e66e1dee3e8e005106fb6a68",
    "pre_install": "Invoke-ExternalCommand -Path \"$(Get-HelperPath -Helper 7zip)\" -ArgumentList @('x', \"-t# `\"$dir\\msmpisetup.exe`\" -o$dir\\tmp\") | Out-Null",
    "architecture": {
        "64bit": {
            "installer": {
                "script": [
                    "Expand-7zipArchive \"$dir\\tmp\\4.msi\" \"$dir\"",
                    "Remove-Item \"$dir\\msmpires.dll\"",
                    "Remove-Item \"$dir\\msmpi.dll\"",
                    "Rename-Item \"$dir\\msmpires64.dll\" \"$dir\\msmpires.dll\"",
                    "Rename-Item \"$dir\\msmpi64.dll\" \"$dir\\msmpi.dll\""
                ]
            }
        },
        "32bit": {
            "installer": {
                "script": "Expand-7zipArchive \"$dir\\tmp\\2.msi\" \"$dir\""
            }
        }
    },
    "post_install": [
        "Remove-Item -Recurse \"$dir\\tmp\"",
        "Remove-Item \"$dir\\msmpisetup.exe\""
    ],
    "env_add_path": ".",
    "checkver": {
        "github": "https://github.com/Microsoft/Microsoft-MPI"
    },
    "autoupdate": {
        "url": "https://github.com/Microsoft/Microsoft-MPI/releases/download/v$version/msmpisetup.exe"
    }
}
```

</details>

<details>
<summary>ida-free</summary>

```json
{
    "version": "8.3",
    "description": "A multi-processor disassembler and debugger that offers so many features it is hard to describe them all",
    "homepage": "https://hex-rays.com/ida-free/",
    "license": "Freeware",
    "architecture": {
        "64bit": {
            "url": "https://out7.hex-rays.com/files/idafree83_windows.exe",
            "hash": "sha1:a63fe3107846ac60fd053c4312482f79d8ab179a"
        }
    },
    "pre_install": "if (!(is_admin)) { throw 'Administrator privileges are required' }",
    "installer": {
        "args": [
            "--mode unattended",
            "--prefix $dir"
        ]
    },
    "bin": "ida64.exe",
    "shortcuts": [
        [
            "ida64.exe",
            "IDA Freeware"
        ]
    ],
    "pre_uninstall": "if (!(is_admin)) { throw 'Administrator privileges are required' }",
    "uninstaller": {
        "file": "uninstall.exe",
        "args": [
            "--mode",
            "unattended"
        ]
    },
    "checkver": {
        "url": "https://hex-rays.com/ida-free/#download",
        "regex": "IDA\\sv([\\d.]+)\\s+"
    },
    "autoupdate": {
        "architecture": {
            "64bit": {
                "url": "https://out7.hex-rays.com/files/idafree$majorVersion$minorVersion_windows.exe",
                "hash": {
                    "url": "https://hex-rays.com/ida-free/#download",
                    "regex": "$sha1\\s+$basename"
                }
            }
        }
    }
}
```

</details>

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have ensured that I am targeting the `develop` branch.
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.
- [x] I have added an entry in the CHANGELOG.

### CODE REVIEW

**Comment by chawyehsu** [comment]
A patch addressing the critical #5065 finally! I'll take a serious review on it. I thought you'd address it in your work of #5124 but this one sounds better to me atm.

**Comment by niheaven** [comment]
I initially planned to address it in #5124, but upon realizing that the PR was a refactoring, I decided to make it more manageable for review by separating the fix.

**Comment by chawyehsu** [comment]
I have tested it with almost all manifests broken by #5065 , including cygwin (dash format args) and miniconda (slash format args), this patch _works_ in both PS5 and PS7. Those two manifests are still problematic - _cygwin shows GUI and miniconda won't install to directory having a dash symbol_, but that's nothing wrong with this PR, args were parsed all correctly.

I don't see any existing app in offical buckets having an installer using args not starting with dash or slash. The regex is well crafted!

File: lib/core.ps1
Comment (chawyehsu) [inline lib/core.ps1:671]: Do we really need to quote the arg when there is white space in it?

Though the regex introduced above parses args correctly. Quoting the arg can be problematic in some cases and the arguable one could be the `/D` arg of NSIS installers. While using `/D=` to specify install location of a NSIS installer, the path may have white spaces. In this case the arg is quote with double quotes, however, NSIS installer will not use such quoted arg and cause unexpected behaviors - custom install location does not work and software is installed to the default location specified by the installer.

test case:
```sh
# Arg /D is not used by the installer
any-nsis-installer.exe /S "/D=X:\Users\John Wick\scoop\apps\any-nsis\versioned"

# Arg /D is used
any-nsis-installer.exe /S /D=X:\Users\John Wick\scoop\apps\any-nsis\versioned
```

Comment (niheaven) [reply lib/core.ps1:671]: I could use regex to just quote the path. Since in Windows, the path must start with `X:` (full path), at least in Scoop, right?

Comment (chawyehsu) [reply lib/core.ps1:671]: According to the [NSIS installer command line usage doc](https://nsis.sourceforge.io/Docs/Chapter3.html), no quote should exist in the arg, therefore quoting the path does not seem to work, I haven't tested it though.

Comment (niheaven) [reply lib/core.ps1:671]: Hmm, yes, no quotes should exist here, I've tested by `7tt`

Comment (niheaven) [reply lib/core.ps1:671]: But I'm still unsure if spaces without quotes will be recognized correctly by some programs.

Comment (chawyehsu) [reply lib/core.ps1]: Maybe an exception can be made for NSIS installer when it's detected? And quotes are still applied in other cases.

Comment (niheaven) [reply lib/core.ps1]: Done

File: lib/core.ps1
Comment (chawyehsu) [inline lib/core.ps1:669]: This doesn't seem to work in PS5. Test manifest:

cygwin.json
```json
{
    "version": "2.931",
    "description": "A large collection of GNU and Open Source tools which provide functionality similar to a Linux distribution on Windows",
    "homepage": "https://cygwin.com/",
    "license": {
        "identifier": "GPL-3.0-or-later,X11,Public Domain",
        "url": "https://cygwin.com/faq.html#faq.what.free"
    },
    "notes": "To start a Cygwin shell, type \"cygwin\".",
    "architecture": {
        "64bit": {
            "url": "https://cygwin.com/setup-x86_64.exe#/cygwin-setup.exe",
            "hash": "9aeacdcf8ff71e2d1a1f9b4685eac12c4de4bb312d44ddac2c7b226f5952bfde"
        },
        "32bit": {
            "url": "https://cygwin.com/setup-x86.exe#/cygwin-setup.exe",
            "hash": "2f1fb00d2693c431d6878799fce00e86a1a238afb8fd4289590708861fb27d8a"
        }
    },
    "installer": {
        "args": [
            "--no-admin",
            "--no-shortcuts",
            "--quiet-mode",
            "--local-package-dir",
            "C:\\Tmp\\John Wick\\packages",
            "--root",
            "C:\\Tmp\\John Wick\\root",
            "--site",
            "https://mirrors.ustc.edu.cn/cygwin/"
        ],
        "keep": true
    },
    "bin": [
        [
            "root\\bin\\bash.exe",
            "cygwin",
            "--login -i"
        ]
    ],
    "shortcuts": [
        [
            "root\\Cygwin.bat",
            "Cygwin",
            "",
            "root\\Cygwin-Terminal.ico"
        ],
        [
            "cygwin-setup.exe",
            "Cygwin Setup",
            "--no-admin --no-shortcuts --root \"$persist_dir\\root\" --local-package-dir \"$persist_dir\\packages\""
        ],
        [
            "root\\bin\\mintty.exe",
            "Cygwin Terminal",
            "-i \"$dir\\root\\Cygwin-Terminal.ico\" -",
            "root\\Cygwin-Terminal.ico"
        ]
    ],
    "checkver": {
        "url": "https://cygwin.com/git/?p=cygwin-apps/setup.git;a=tags",
        "regex": ">release_([\\d.]+)<"
    },
    "autoupdate": {
        "architecture": {
            "64bit": {
                "url": "https://cygwin.com/setup-x86_64.exe#/cygwin-setup.exe"
            },
            "32bit": {
                "url": "https://cygwin.com/setup-x86.exe#/cygwin-setup.exe"
            }
        }
    }
}
```

The path `C:\\Tmp\\John Wick\\` is a simulation.

Comment (niheaven) [reply lib/core.ps1]: OK, in PS5 the assign statement doesn't return value, and I made a stupid mistake.

**Review by chawyehsu — APPROVED** [review]
I don't have any questions now, thank you for your work on this @niheaven.

Imo this is a critical part of the core therefore I suggest one more reviewer on this, but I don't know if any member is currently available. Please review when you are available. Good luck! /cc @ScoopInstaller/maintainers

**Comment by niheaven** [comment]
I will merge this to develop after 24 hours for further testing. This seems to be the more appropriate approach.

**Pull request merged**
